### PR TITLE
Extended PyTorchFileLoader class.

### PR DESCRIPTION
**Pull request by putivsky**
Summary: Added Fusion JIT Graph loading into Glow Function.

Differential Revision: D16678267

### CODE REVIEW

File: torch_glow/src/PyTorchFileLoader.cpp [hunk 1]
```diff
@@ -15,30 +15,188 @@
...
+            return 0;
+          };
+        },
+        options)});
+
+    torch::jit::RegisterPass pass([&](std::shared_ptr<torch::jit::Graph> &g) {
+      // Trigger custom fusion pass only if local thread Glow Function is set.
+      if (localFusionInfo.function) {
+        // Fuse all linear operators.
+        FuseLinear(g);
```
Comment (bwasti) [inline torch_glow/src/PyTorchFileLoader.cpp]: I guess backends would generally benefit from these canonicalization passes living in Pytorch's core, huh? @kimishpatel this is what we discussed in our meeting

File: torch_glow/src/PyTorchFileLoader.cpp [hunk 2]
```diff
@@ -15,30 +15,188 @@
...
+/*static*/
+llvm::Error PyTorchFileLoader::loadPyTorchGraph(
+    const std::string &fileName, std::vector<torch::jit::IValue> &inputs,
+    glow::Function &F, std::vector<glow::Placeholder *> &inputPlaceholders,
+    std::vector<glow::Placeholder *> &outputPlaceholders, bool sanityCheck) {
+  // Register custom pass once.
+  static RegisterCustomFusionPass fuser;
+
+  // Convert PyTorch model into JIT Module.
+  std::shared_ptr<torch::jit::script::Module> module = {nullptr};
```
Comment (bwasti) [inline torch_glow/src/PyTorchFileLoader.cpp]: nit: do you need to initialize a shared pointer like this? I think the default constructor is empty

File: torch_glow/src/PyTorchFileLoader.cpp (hunk 1 above)
Comment (putivsky) [reply torch_glow/src/PyTorchFileLoader.cpp]: Yeah, I have thought about moving pass into "common" file, need an agreement from code owner first.

Comment (putivsky) [reply torch_glow/src/PyTorchFileLoader.cpp]: Moved.

File: torch_glow/src/FusingOptimizer.cpp [hunk 3]
```diff
@@ -89,4 +92,16 @@ graph(%input, %weight):
   mm_to_linear.runOnGraph(graph);
 }
 
+void glowCustomFuse(std::shared_ptr<torch::jit::Graph> &g,
```
Comment (jackm321) [inline torch_glow/src/FusingOptimizer.cpp]: We should eventually split fusions here into different files, can this go into PyTorchCommon.cpp?

File: torch_glow/src/PyTorchFileLoader.cpp [hunk 4]
```diff
@@ -15,30 +15,182 @@
...
+        getFusionSymbol(),
+        [](const torch::jit::Node *node) {
+          return [node](torch::jit::Stack &stack) {
+            // Get JIT Graph.
+            auto graph = node->g(at::attr::Subgraph);
+            auto err = loadJitGraphToGlowFunction(
+                stack, *graph, *localFusionInfo.function,
+                *localFusionInfo.inputPlaceholders,
+                *localFusionInfo.outputPlaceholders);
+            if (static_cast<bool>(err)) {
```
Comment (jackm321) [inline torch_glow/src/PyTorchFileLoader.cpp]: glow errToBool

File: torch_glow/src/PyTorchFileLoader.cpp [hunk 5]
```diff
@@ -15,30 +15,182 @@
...
+  at::NoGradGuard guard;
+
+  // Set thread local pointer, not null values activate custom fusion pass.
+  localFusionInfo.set(&F, &inputPlaceholders, &outputPlaceholders);
+  auto err = evaluateModuleGraph(module, inputs);
+  localFusionInfo.reset();
+
+  RETURN_IF_ERR(err);
+
+  if (sanityCheck) {
```
Comment (jackm321) [inline torch_glow/src/PyTorchFileLoader.cpp]: could honestly be in another function but not a big deal

File: torch_glow/src/PyTorchFileLoader.h [hunk 6]
```diff
@@ -26,10 +27,20 @@ namespace glow {
...
-                    std::shared_ptr<torch::jit::script::Module> &module,
-                    llvm::Error *errPtr);
+  /// \returns error if any.
+  static llvm::Error
+  loadPyTorchModel(const std::string &fileName,
+                   std::shared_ptr<torch::jit::script::Module> &module);
+
+  /// Takes a model file \p fileName, loads optimized graph into Glow Function
+  /// \p F and input \p inputPlaceholders, output \p outputPlaceholders
+  /// placeholders, \returns error if any.
```
Comment (jackm321) [inline torch_glow/src/PyTorchFileLoader.h]: Can you add some comments about the thread_local stuff here. How should this be used to take advantage of that? For example, "call this once per inference thread for a model" or something like that

File: torch_glow/src/PyTorchLoaderTest.cpp [hunk 7]
```diff
@@ -16,149 +16,45 @@
...
+  llvm::Error err = glow::PyTorchFileLoader::loadPyTorchGraph(
+      fileName, vec, *F, inputPlaceholders, outputPlaceholders);
+
+  if (static_cast<bool>(err)) {
+    LOG(ERROR) << llvm::toString(std::move(err)) << "\n"; // size and reshape.
+  }
+}
+} // namespace
+
+int main(int /*argc*/, char ** /*argv*/) {
```
[the remark this answers is not in the document]
Comment (putivsky) [reply torch_glow/src/PyTorchLoaderTest.cpp]: I wan't able to build gtest under github/CMakefile.

File: torch_glow/src/PyTorchFileLoader.cpp (hunk 4 above)
Comment (putivsky) [reply torch_glow/src/PyTorchFileLoader.cpp]: I don't want to output error, instead just to throw exception.

File: torch_glow/src/PyTorchLoaderTest.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (jackm321) [reply torch_glow/src/PyTorchLoaderTest.cpp]: Hmm ok well all of our other c++ tests use gtest with cmake, I think we should continue using it here

File: torch_glow/src/PyTorchFileLoader.cpp (hunk 4 above)
Comment (jackm321) [reply torch_glow/src/PyTorchFileLoader.cpp]: I see ok

File: torch_glow/src/PyTorchLoaderTest.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (putivsky) [reply torch_glow/src/PyTorchLoaderTest.cpp]: I have meant:
python setup.py test --run_cmake
failed.

[the remark this answers is not in the document]
Comment (jackm321) [reply torch_glow/src/PyTorchLoaderTest.cpp]: What's the error?

[the remark this answers is not in the document]
Comment (putivsky) [reply torch_glow/src/PyTorchLoaderTest.cpp]: Something like, cannot find file: #include "gtest/gtest.h"

[the remark this answers is not in the document]
Comment (jackm321) [reply torch_glow/src/PyTorchLoaderTest.cpp]: And you included gtest in target_link_libraries?

[the remark this answers is not in the document]
Comment (putivsky) [reply torch_glow/src/PyTorchLoaderTest.cpp]: yes, I have added gtest to CMakefile.

[the remark this answers is not in the document]
Comment (jackm321) [reply torch_glow/src/PyTorchLoaderTest.cpp]: 😞 oh well let's move no as is then.
It would be great if this was like the rest of our tests. Integration with pytorch jit via C++  is important and we should add some more tests here eventaully, we can follow up on trying to get gtest working then

File: torch_glow/src/PyTorchLoaderTest.cpp [hunk 8]
```diff
@@ -16,151 +16,44 @@
...
-    auto ptT = at::empty(sizes);
-    auto var = torch::autograd::make_variable(ptT);
-    // Add to stack output parameter.
-    stack.push_back(at::IValue(var));
-  }
-}
-} // namespace
-
-TEST(PyTorchModelLoader, Model) {
+void pyTorchModelLoaderModelTest() {
```
Comment (jackm321) [inline torch_glow/src/PyTorchLoaderTest.cpp]: Can you move this to the tests directory

File: torch_glow/src/PyTorchModelLoader.h [hunk 9]
```diff
@@ -53,7 +53,7 @@ class PyTorchModelLoader {
         llvm::Error (PyTorchModelLoader::*)(const torch::jit::Node *);
 
     /// Symbols (as strings) that this mapping value is applicable to.
-    const std::vector<const char *> symbols;
+    std::vector<const char *> symbols;
```
Comment (jackm321) [inline torch_glow/src/PyTorchModelLoader.h]: why this change?

File: torch_glow/src/PyTorchLoaderTest.cpp (hunk 8 above)
Comment (jackm321) [reply torch_glow/src/PyTorchLoaderTest.cpp]: doxygen

File: torch_glow/src/PyTorchFileLoader.h [hunk 10]
```diff
@@ -17,19 +17,35 @@
...
+  /// Method is thread safe, internally it uses local thread structures for
+  /// executing custom fusion pass, registered globally. No other passes or
+  /// other treads calling this method will be affected.
+  static llvm::Error loadPyTorchGraph(
```
Comment (jackm321) [inline torch_glow/src/PyTorchFileLoader.h]: doxygen like for example "\p sanityCheck ..."

File: torch_glow/src/PyTorchFileLoader.cpp [hunk 11]
```diff
@@ -15,30 +15,189 @@
...
+  auto err = evaluateModuleGraph(module, inputs);
+  localFusionInfo.reset();
+
+  RETURN_IF_ERR(err);
+
+  return sanityCheck ? performSanityCheck() : llvm::Error::success();
+}
+
+/*static*/
+llvm::Error PyTorchFileLoader::performSanityCheck() {
```
Comment (jackm321) [inline torch_glow/src/PyTorchFileLoader.cpp]: thanks for moving this

File: torch_glow/src/PyTorchFileLoader.cpp [hunk 12]
```diff
@@ -15,30 +15,189 @@
...
+  // Sanity check, after "fusionSymbol" node, not other nodes should exist.
```
Comment (jackm321) [inline torch_glow/src/PyTorchFileLoader.cpp]: move this comment above the function

File: torch_glow/src/PyTorchFileLoader.cpp [hunk 13]
```diff
@@ -15,30 +15,189 @@
...
+/*static*/
+llvm::Error PyTorchFileLoader::loadPyTorchGraph(
+    const std::string &fileName, std::vector<torch::jit::IValue> &inputs,
+    glow::Function &F, std::vector<glow::Placeholder *> &inputPlaceholders,
+    std::vector<glow::Placeholder *> &outputPlaceholders,
+    const PyTorchLoaderSettings &settings, bool sanityCheck) {
+  // Register custom pass once.
+  static RegisterCustomFusionPass fuser;
+
+  // Convert PyTorch model into JIT Module.
```
Comment (jackm321) [inline torch_glow/src/PyTorchFileLoader.cpp]: doxygen

File: torch_glow/src/PyTorchFileLoader.cpp [hunk 14]
```diff
@@ -15,30 +15,189 @@
...
+  std::shared_ptr<torch::jit::script::Module> module;
+  RETURN_IF_ERR(PyTorchFileLoader::loadPyTorchModel(fileName, module));
+
+  // Disable gradient nodes generation.
+  at::NoGradGuard guard;
+
+  // Set thread local pointer, not null values activate custom fusion pass.
```
Comment (jackm321) [inline torch_glow/src/PyTorchFileLoader.cpp]: doxygen

File: torch_glow/src/PyTorchCommon.h [hunk 15]
```diff
@@ -40,6 +40,11 @@ PyTorchLoaderSettings &getPyTorchLoaderSettings();
 /// \returns the PyTorch symbol to be used for the PyTorch node which represents
 /// the subgraph that Glow will compile and run.
 const c10::Symbol &getGlowSymbol();
+
+/// Executes custom fuse pass for the given \p graph and \p fuseSymbol.
+void glowCustomFuse(std::shared_ptr<torch::jit::Graph> &graph,
```
Comment (jackm321) [inline torch_glow/src/PyTorchCommon.h]: where is the thread_local stuff documented?

File: torch_glow/src/PyTorchFileLoader.cpp [hunk 16]
```diff
@@ -15,30 +15,189 @@
...
+    inputPlaceholders = in;
+    outputPlaceholders = out;
+    settings = s;
+  }
+
+  void reset() { set(nullptr, nullptr, nullptr, nullptr); }
+};
+
+/// Meyer's singleton for static fusion symbol.
+static at::Symbol getFusionSymbol() {
```
Comment (jackm321) [inline torch_glow/src/PyTorchFileLoader.cpp]: why do we need another symbol?

File: torch_glow/src/PyTorchFileLoader.cpp [hunk 17]
```diff
@@ -15,30 +15,189 @@
...
+llvm::Error PyTorchFileLoader::performSanityCheck() {
+  // Sanity check, after "fusionSymbol" node, not other nodes should exist.
+  std::shared_ptr<torch::jit::Graph> subgraph =
+      torch::jit::lastExecutedOptimizedGraph();
+  size_t fusedNodes = 0, missedNodes = 0;
+  const auto symbol = getFusionSymbol();
+  for (const auto &node : subgraph->nodes()) {
+    if (node->kind() == symbol) {
+      ++fusedNodes;
+    } else if (fusedNodes) {
```
Comment (jackm321) [inline torch_glow/src/PyTorchFileLoader.cpp]: This will only work if the non-fused nodes are after the fusion group. Is this the intention? if so, why?

File: torch_glow/src/PyTorchModelLoader.h (hunk 9 above)
Comment (putivsky) [reply torch_glow/src/PyTorchModelLoader.h]: By design initialization of the members must be done at construction time, this is self-check making sure no one can change members after initialization.

File: torch_glow/src/PyTorchCommon.h (hunk 15 above)
Comment (putivsky) [reply torch_glow/src/PyTorchCommon.h]: Nowhere, this is hidden from public implementation details, caller shoudn't even know about all this local thread stuff. What caller needs to know is the ability to invoke public methods of PyTorchFileLoader class in multiple threads.

File: torch_glow/src/PyTorchFileLoader.cpp (hunk 16 above)
Comment (putivsky) [reply torch_glow/src/PyTorchFileLoader.cpp]: Yes, because we a registering different operators and correspondently different fuse passes.

File: torch_glow/src/PyTorchFileLoader.cpp (hunk 17 above)
Comment (putivsky) [reply torch_glow/src/PyTorchFileLoader.cpp]: Normally - ALL operators must be fused into one graph node. If you take a look at graph it contains from the beginning the original graph nodes and then fused node, even all nodes got fused.

File: torch_glow/src/PyTorchFileLoader.cpp (hunk 14 above)
Comment (putivsky) [reply torch_glow/src/PyTorchFileLoader.cpp]: can you explain, what should be changed to comply with doxygen style?.

File: torch_glow/src/PyTorchCommon.h (hunk 15 above)
Comment (jackm321) [reply torch_glow/src/PyTorchCommon.h]: discussed offline, makes sense (I think)

File: torch_glow/src/PyTorchFileLoader.cpp (hunk 17 above)
Comment (jackm321) [reply torch_glow/src/PyTorchFileLoader.cpp]: Oh whoa, that seems super weird. I think I see why that's happening, but why is that desirable instead of deleting the PT graph nodes once they've been fused into Glow?

File: torch_glow/setup.py [hunk 18]
```diff
@@ -124,6 +125,7 @@ def _run_cmake(self):
                 # PyTorch cmake args
                 '-DPYTORCH_DIR={}'.format(
                     os.path.dirname(os.path.realpath(torch.__file__))),
+                '-DTORCH_GLOW={}'.format(FILE_DIR),
```
Comment (jackm321) [inline torch_glow/setup.py]: What is this?

Comment (putivsky) [reply torch_glow/setup.py]: Normally we should use include paths relative to "glow" folder, otherwise we would need (as in this case) path to the include file from different folder. Because I have moved test file into a separate folder, test makefile needs include path to other torch_glow files.
